torch/csrc/jit/graph_executor.cpp

```diff
@@ -0,0 +1,396 @@
+  // capture list because the list is not sorted with inputs before outputs.
+  // We cannot capture both, because we actually destroy our input lists (to release references)
+  // before we get the outputs list.
+  void captureInputs(ExecutionPlanAutogradFunction & grad_fn, variable_tensor_list & inputs) {
+    auto & capture_desc = grad.df_input_captures;
+    size_t N = capture_desc.size();
+    for(size_t i = 0; i < N; ++i) {
+      if(capture_desc[i].kind == Capture::Kind::Input) {
+        size_t offset = capture_desc[i].offset;
+        grad_fn.captures[i] = autograd::SavedVariable(autograd::Variable(inputs[offset]), false);
+      }
+    }
+  }
+  void captureOutputs(ExecutionPlanAutogradFunction & grad_fn, variable_tensor_list & outputs) {
+    auto & capture_desc = grad.df_input_captures;
+    size_t N = capture_desc.size();
+    for(size_t i = 0; i < N; ++i) {
+      if(capture_desc[i].kind == Capture::Kind::Output) {
+        size_t offset = capture_desc[i].offset;
+        grad_fn.captures[i] = autograd::SavedVariable(autograd::Variable(outputs[offset]), true);
+      }
+    }
+  }
+
+  variable_tensor_list runWithGrad(variable_tensor_list inputs) {
+    auto grad_fn = std::make_shared<ExecutionPlanAutogradFunction>(grad_executor, grad.df_input_captures.size());
+    // hook up the outputs of df to the gradient functions of the inputs that require
+    // gradients
+    for(auto idx : grad.df_output_vjps) {
+      autograd::Variable v(inputs[idx]);
+      // TODO: this kinda stuff is _way_ to low level to the public API of variable.
+      // Why do I have to care here whether v has a grad_fn or grad accumulator?
+      // Why do I have to care here about output_nr? I just want to say
+      // grad_fn->setOutputTo(i, v.input_port());
+      grad_fn->next_functions.emplace_back(v.grad_fn() ? v.grad_fn() : v.grad_accumulator(), v.output_nr());
+    }
+    captureInputs(*grad_fn, inputs);
+
+    unwrapVariables(inputs);
+    tensor_list outputs;
+    InterpreterState(f).runOneStage(std::move(inputs), outputs);
+    wrapTensors(outputs);
+    // hookup the gradients for the output tensors that require gradients
+    // to the inputs to our gradient function df
+    // TODO - XXX - if any output is the same tensor multiple times, views have to be
+    // setup here. We need to refactor autograd until it is safe for
+    // tensors to be constructed without all the viewing infrastructure.
+    // this is currently intentionally not done here so we can get an idea of our
+    // perf before introducing overhead for correctness
+    for(auto idx : grad.df_input_vjps) {
+      autograd::Variable o(outputs[idx]);
+      auto impl = o.get();
+      // Note: we have to set this up in place, or we have to
+      // throw away and reallocate variables that were already created in
+      // wrapTensors. We should add an API for this, and more generally
+      // we need to clean up the fields of Variable.
```

Comment:
Sorry, why would we have to reallocate Variables if we used `set_history` instead of fiddling with internals? This is unsafe

Reply:
Neither of the `set_history` functions do the same thing as this. One sets `output_nr` to zero. The other expects a contiguous list of tensors which we do not have. So we would need another function. Clarify the variable API would help move these implementation details into the autograd.

Reply:
Does it make sense to just add another variant of set_history? We are guaranteed to have just allocated these Variables to be non-gradient things that are otherwise do not have grad information setup. But the way the gradient meta-data is setup we only discover we need to hook up a gradient but iterating a the df_input_vjps list.

Reply:
Hmm I thought there's a `set_history` you can call on a Variable, and you can pass it all those details, but I might be mistaken.